### ISSUE TITLE
Noisy, non-actionable, exception thrown when exiting continuous build

### ISSUE THREAD

**Issue by alkemist** [gradle/gradle#2224]
For an internal project, we reliably get the following when exiting continuous build:

```
java.lang.IllegalStateException: Cannot dispatch message, as /0:0:0:0:0:0:0:1:54407 to /0:0:0:0:0:0:0:1:54406 has been stopped.
        at org.gradle.internal.remote.internal.hub.MessageHub.assertRunning(MessageHub.java:168)
        at org.gradle.internal.remote.internal.hub.MessageHub.access$600(MessageHub.java:44)
        at org.gradle.internal.remote.internal.hub.MessageHub$ChannelDispatch.dispatch(MessageHub.java:363)
        at org.gradle.internal.dispatch.ProxyDispatchAdapter$DispatchingInvocationHandler.invoke(ProxyDispatchAdapter.java:93)
        at com.sun.proxy.$Proxy0.sendOutputEvent(Unknown Source)
        at org.gradle.process.internal.worker.child.WorkerLogEventListener.onOutput(WorkerLogEventListener.java:36)
        at sun.reflect.GeneratedMethodAccessor5.invoke(Unknown Source)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:498)
        at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:35)
        at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:24)
        at org.gradle.internal.event.AbstractBroadcastDispatch.dispatch(AbstractBroadcastDispatch.java:42)
        at org.gradle.internal.event.BroadcastDispatch$SingletonDispatch.dispatch(BroadcastDispatch.java:230)
        at org.gradle.internal.event.BroadcastDispatch$SingletonDispatch.dispatch(BroadcastDispatch.java:149)
        at org.gradle.internal.event.AbstractBroadcastDispatch.dispatch(AbstractBroadcastDispatch.java:58)
        at org.gradle.internal.event.BroadcastDispatch$CompositeDispatch.dispatch(BroadcastDispatch.java:324)
        at org.gradle.internal.event.BroadcastDispatch$CompositeDispatch.dispatch(BroadcastDispatch.java:234)
        at org.gradle.internal.event.ListenerBroadcast.dispatch(ListenerBroadcast.java:140)
        at org.gradle.internal.event.ListenerBroadcast.dispatch(ListenerBroadcast.java:37)
        at org.gradle.internal.dispatch.ProxyDispatchAdapter$DispatchingInvocationHandler.invoke(ProxyDispatchAdapter.java:93)
        at com.sun.proxy.$Proxy21.onOutput(Unknown Source)
        at org.gradle.internal.logging.sink.OutputEventRenderer.onOutput(OutputEventRenderer.java:334)
        at org.gradle.internal.logging.services.TextStreamOutputEventListener.onTextEvent(TextStreamOutputEventListener.java:57)
        at org.gradle.internal.logging.services.TextStreamOutputEventListener.onOutput(TextStreamOutputEventListener.java:41)
        at org.gradle.internal.logging.source.PrintStreamLoggingSystem$OutputEventDestination.onOutput(PrintStreamLoggingSystem.java:165)
        at org.gradle.internal.logging.source.PrintStreamLoggingSystem$1.text(PrintStreamLoggingSystem.java:43)
        at org.gradle.internal.io.LineBufferingOutputStream.flush(LineBufferingOutputStream.java:95)
        at org.gradle.internal.io.LineBufferingOutputStream.write(LineBufferingOutputStream.java:80)
        at java.io.OutputStream.write(OutputStream.java:116)
        at java.io.PrintStream.write(PrintStream.java:480)
        at sun.nio.cs.StreamEncoder.writeBytes(StreamEncoder.java:221)
        at sun.nio.cs.StreamEncoder.implFlushBuffer(StreamEncoder.java:291)
        at sun.nio.cs.StreamEncoder.flushBuffer(StreamEncoder.java:104)
        at java.io.OutputStreamWriter.flushBuffer(OutputStreamWriter.java:185)
        at java.io.PrintStream.newLine(PrintStream.java:546)
        at java.io.PrintStream.println(PrintStream.java:824)
        at org.gradle.internal.io.LinePerThreadBufferingOutputStream.println(LinePerThreadBufferingOutputStream.java:191)
        at java.lang.Throwable$WrappedPrintStream.println(Throwable.java:748)
        at java.lang.Throwable.printStackTrace(Throwable.java:655)
        at java.lang.Throwable.printStackTrace(Throwable.java:643)
        at org.gradle.internal.logging.slf4j.OutputEventListenerBackedLogger.log(OutputEventListenerBackedLogger.java:138)
        at org.gradle.internal.logging.slf4j.OutputEventListenerBackedLogger.error(OutputEventListenerBackedLogger.java:445)
        at org.gradle.internal.remote.internal.hub.MessageHubBackedObjectConnection$1.execute(MessageHubBackedObjectConnection.java:59)
        at org.gradle.internal.remote.internal.hub.MessageHubBackedObjectConnection$1.execute(MessageHubBackedObjectConnection.java:57)
        at org.gradle.internal.remote.internal.hub.MessageHub$Handler.run(MessageHub.java:426)
        at org.gradle.internal.concurrent.ExecutorPolicy$CatchAndRecordFailures.onExecute(ExecutorPolicy.java:63)
        at org.gradle.internal.concurrent.StoppableExecutorImpl$1.run(StoppableExecutorImpl.java:46)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
        at org.gradle.internal.concurrent.ThreadFactoryImpl$ManagedThreadRunnable.run(ThreadFactoryImpl.java:55)
        at java.lang.Thread.run(Thread.java:745)
```

I'm also not entirely sure when this started happening, but I _think_ it started happening with 4.0.

The build launches a worker/deployment. My guess is that when we tell the deployment to shut down, it logs something, which the worker infrastructure tries to send back to the Gradle process but we've already severed the connection.

**Comment by eriwen** [comment]
@alkemist Could you please consider sharing project details via email or a private channel?

**Comment by ghale** [comment]
@alkemist is probably correct.  It looks like we stop the connection _before_ we remove the output event listener from the logging manager in worker processes (see https://github.com/gradle/gradle/blob/9a6d0e8621b8fcbfe8c6f178cd25d3c1d21e67af/subprojects/core/src/main/java/org/gradle/process/internal/worker/child/SystemApplicationClassLoaderWorker.java#L121-L121).

I think the sequence here should be:

1. Remove output listener
1. Stop connection
1. Stop messaging services
1. Stop logging manager

Digging back, this likely affected at least 3.5, possibly as early as 3.3.

**Comment by ghale** [comment]
It does seem odd that the message is being sent in between when the connection is stopped and the output event listener is removed, though.  Perhaps there is some new buffering in the logging manager that is causing the delayed message, which is why this is just showing up in 4.0.  Regardless, fixing the "stop" order will likely eliminate the failure.

**Comment by alkemist** [comment]
One thing to note is that this logging from the worker is happening during a JVM shutdown hook callback. 

At worst, we just need to not thrown an exception in this circumstance. At best, we somehow don't stop listening for output until the worker process has _really_ shutdown.

**Comment by ghale** [comment]
Oh, that's something else then.  A JVM shutdown hook should not be logging through the output event listeners at all.  The whole point of removing the output listener and stopping the logging manager at the end of the worker is to reset stdout/stderr so that we don't try to use these mechanisms after we've given up control of execution.  It's likely that something regressed in the logging framework that is keeping this reset from working properly.

**Comment by ghale** [comment]
Do we _know_ it's the output from the shutdown hook that is causing the failure?  We do have some unit test coverage for this scenario (output on shutdown hook) and I tried this with some integration tests and it seems to work as expected.

Digging a little deeper, the `MessageHub` has an async queue of messages that he works through, sending to the various dispatches that he has registered.  It's possible that there is a backlog of messages at the time the connection is closed, causing this error.  Looking at the stacktrace, it seems to fit that hypothesis.

**Comment by ghale** [comment]
Actually, it looks like the `MessageHub` gets stopped with the connection , so it's not a backlog there.  Maybe it's some sort of async output from the logging manager after all.

**Comment by ghale** [comment]
@lacasseio and I chatted about this a bit today and we have a few theories of what _could_ be happening, but I have yet to reproduce the behavior.  I can easily fix the known issue of the stop ordering, but I would prefer to be able to capture the issue with a regression test.  @alkemist is the code for the worker/deployment that is causing this somewhere you can share with me offline?

**Comment by alkemist** [comment]
@gary I don't actually know anything here, it's just guesswork on my part.

**Comment by ghale** [comment]
Ok, after some digging, the issue is because there's another thread that's producing output while the worker is shutting down.  This causes an error when the connection is shutdown and the outputEventListener is still registered.  I've confirmed that re-ordering the shutdown sequence prevents the exception from occurring.

See https://github.com/gradle/gradle/pull/2244.

**Comment by ghale** [comment]
This should be fixed in 4.0-rc-2.

**Comment by alkemist** [comment]
Awesome. Thanks @ghale.
